"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var chai_1 = require("chai");
var sinon = require("sinon");
var subscribe_1 = require("../src/subscribe");
describe('createSubscribe', function () ***REMOVED***
    var spy;
    beforeEach(function () ***REMOVED***
        // Listen to console.error calls.
        spy = sinon.spy(console, 'error');
    ***REMOVED***);
    afterEach(function () ***REMOVED***
        spy.restore();
    ***REMOVED***);
    it('Creation', function (done) ***REMOVED***
        var subscribe = subscribe_1.createSubscribe(function (observer) ***REMOVED***
            observer.next(123);
        ***REMOVED***);
        var unsub = subscribe(function (value) ***REMOVED***
            unsub();
            chai_1.assert.equal(value, 123);
            done();
        ***REMOVED***);
    ***REMOVED***);
    it('Logging observer error to console', function (done) ***REMOVED***
        var uncatchableError = new Error('uncatchable');
        var subscribe = subscribe_1.createSubscribe(function (observer) ***REMOVED***
            observer.next(123);
            observer.complete();
        ***REMOVED***);
        subscribe(***REMOVED***
            next: function (value) ***REMOVED***
                chai_1.assert.equal(value, 123);
                // Simulate an error is thrown in the next callback.
                // This should log to the console as an error.
                throw uncatchableError;
            ***REMOVED***,
            complete: function () ***REMOVED***
                // By this point, the error should have been logged.
                chai_1.assert.ok(spy.calledWith(uncatchableError));
                done();
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***);
    it('Well-defined subscription order', function (done) ***REMOVED***
        var subscribe = subscribe_1.createSubscribe(function (observer) ***REMOVED***
            observer.next(123);
            // Subscription after value emitted should NOT be received.
            subscribe(***REMOVED***
                next: function (value) ***REMOVED***
                    chai_1.assert.ok(false);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
        // Subscription before value emitted should be recieved.
        subscribe(***REMOVED***
            next: function (value) ***REMOVED***
                done();
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***);
    it('Subscribing to already complete Subscribe', function (done) ***REMOVED***
        var seq = 0;
        var subscribe = subscribe_1.createSubscribe(function (observer) ***REMOVED***
            observer.next(456);
            observer.complete();
        ***REMOVED***);
        subscribe(***REMOVED***
            next: function (value) ***REMOVED***
                chai_1.assert.equal(seq++, 0);
                chai_1.assert.equal(value, 456);
            ***REMOVED***,
            complete: function () ***REMOVED***
                subscribe(***REMOVED***
                    complete: function () ***REMOVED***
                        chai_1.assert.equal(seq++, 1);
                        done();
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***);
    it('Subscribing to errored Subscribe', function (done) ***REMOVED***
        var seq = 0;
        var subscribe = subscribe_1.createSubscribe(function (observer) ***REMOVED***
            observer.next(246);
            observer.error(new Error('failure'));
        ***REMOVED***);
        subscribe(***REMOVED***
            next: function (value) ***REMOVED***
                chai_1.assert.equal(seq++, 0);
                chai_1.assert.equal(value, 246);
            ***REMOVED***,
            error: function (e) ***REMOVED***
                chai_1.assert.equal(seq++, 1);
                subscribe(***REMOVED***
                    error: function (e2) ***REMOVED***
                        chai_1.assert.equal(seq++, 2);
                        chai_1.assert.equal(e.message, 'failure');
                        done();
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***,
            complete: function () ***REMOVED***
                chai_1.assert.ok(false);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***);
    it('Delayed value', function (done) ***REMOVED***
        var subscribe = subscribe_1.createSubscribe(function (observer) ***REMOVED***
            setTimeout(function () ***REMOVED*** return observer.next(123); ***REMOVED***, 10);
        ***REMOVED***);
        subscribe(function (value) ***REMOVED***
            chai_1.assert.equal(value, 123);
            done();
        ***REMOVED***);
    ***REMOVED***);
    it('Executor throws => Error', function () ***REMOVED***
        // It's an application error to throw an exception in the executor -
        // but since it is called asynchronously, our only option is
        // to emit that Error and terminate the Subscribe.
        var subscribe = subscribe_1.createSubscribe(function (observer) ***REMOVED***
            throw new Error('Executor throws');
        ***REMOVED***);
        subscribe(***REMOVED***
            error: function (e) ***REMOVED***
                chai_1.assert.equal(e.message, 'Executor throws');
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***);
    it('Sequence', function (done) ***REMOVED***
        var subscribe = makeCounter(10);
        var j = 1;
        subscribe(***REMOVED***
            next: function (value) ***REMOVED***
                chai_1.assert.equal(value, j++);
            ***REMOVED***,
            complete: function () ***REMOVED***
                chai_1.assert.equal(j, 11);
                done();
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***);
    it('unlisten', function (done) ***REMOVED***
        var subscribe = makeCounter(10);
        subscribe(***REMOVED***
            complete: function () ***REMOVED***
                subscribe_1.async(done)();
            ***REMOVED***
        ***REMOVED***);
        var j = 1;
        var unsub = subscribe(***REMOVED***
            next: function (value) ***REMOVED***
                chai_1.assert.ok(value <= 5);
                chai_1.assert.equal(value, j++);
                if (value === 5) ***REMOVED***
                    unsub();
                ***REMOVED***
            ***REMOVED***,
            complete: function () ***REMOVED***
                chai_1.assert.ok(false, 'Does not call completed if unsubscribed');
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***);
    it('onNoObservers', function (done) ***REMOVED***
        var subscribe = makeCounter(10);
        var j = 1;
        var unsub = subscribe(***REMOVED***
            next: function (value) ***REMOVED***
                chai_1.assert.ok(value <= 5);
                chai_1.assert.equal(value, j++);
                if (value === 5) ***REMOVED***
                    unsub();
                    subscribe_1.async(done)();
                ***REMOVED***
            ***REMOVED***,
            complete: function () ***REMOVED***
                chai_1.assert.ok(false, 'Does not call completed if unsubscribed');
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***);
    // TODO(koss): Add test for partial Observer (missing methods).
    it('Partial Observer', function (done) ***REMOVED***
        var subscribe = makeCounter(10);
        var unsub = subscribe(***REMOVED***
            complete: function () ***REMOVED***
                done();
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***);
function makeCounter(maxCount, ms) ***REMOVED***
    if (ms === void 0) ***REMOVED*** ms = 10; ***REMOVED***
    var id;
    return subscribe_1.createSubscribe(function (observer) ***REMOVED***
        var i = 1;
        id = setInterval(function () ***REMOVED***
            observer.next(i++);
            if (i > maxCount) ***REMOVED***
                if (id) ***REMOVED***
                    clearInterval(id);
                    id = undefined;
                ***REMOVED***
                observer.complete();
            ***REMOVED***
        ***REMOVED***, ms);
    ***REMOVED***, function (observer) ***REMOVED***
        clearInterval(id);
        id = undefined;
    ***REMOVED***);
***REMOVED***

//# sourceMappingURL=subscribe.test.js.map
